### PR TITLE
feat: handle tests for `WATCHER_KUBERNETES` mode

### PR DESCRIPTION
**Pull request by vricciardulli**
## Description

This PR repoens https://github.com/astronomer/astronomer-cosmos/pull/2529 to fix hanging tests.

What was done:
* Support tests in `WATCHER_KUBERNETES` mode.
* Update docs
* Use `CosmosKubernetesPodManager` to provide `context`, `tests_per_model` and `test_results_per_model` instead of using globals.
* Refreshed `manifest.json`

## Checklist

- [x] I have made corresponding changes to the documentation (if required)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: cosmos/operators/watcher_kubernetes.py [hunk 1]
```diff
@@ -172,13 +167,24 @@ def __init__(self, *args: Any, **kwargs: Any):
...
+    aggregated XCom (``"pass"`` or ``"fail"``) under the key returned by
+    ``get_tests_status_xcom_key(model_uid)``. This key sanitizes the dbt
+    model unique ID by replacing ``.`` with ``__`` before appending
+    ``_tests_status`` (for example,
+    ``model.jaffle_shop.stg_orders`` becomes
+    ``model__jaffle_shop__stg_orders_tests_status``).
+
+    This sensor polls that key and returns success when ``"pass"`` or raises
+    ``AirflowException`` when ``"fail"``.
     """
```
[the remark this answers is not in the document]
Comment (vricciardulli) [reply cosmos/operators/watcher_kubernetes.py:228]: looking into it

[the remark this answers is not in the document]
Comment (vricciardulli) [reply cosmos/operators/watcher_kubernetes.py:228]: should be fixed now

**Comment by vricciardulli** [comment]
@tatiana I've made some changes that should fix the tests hanging. I did not run those tests locally though, so whenever you can enable them in the CI we can see if this works, thanks:) (I'm not super happy about the approach of using globals, so let me know if you have any suggestion)

**Comment by tatiana** [comment]
@vricciardulli, thanks a lot for reopening the PR!

Unfortunately, it seems the test is still hanging; this has been running for 4h and normally runs in  ~6m :
Screenshot 2026-04-08 at 11 57 12

We can identify the test that is hanging in the logs:
Screenshot 2026-04-08 at 11 57 34
https://github.com/astronomer/astronomer-cosmos/actions/runs/24129325262/job/70409418823?pr=2543

This is what we'd expect to see if things were working well:
```
tests/test_example_k8s_dags.py::test_example_dag_kubernetes PASSED       [ 33%]
tests/test_example_k8s_dags.py::test_example_dag_watcher_kubernetes PASSED [ 66%]
tests/operators/test_watcher_kubernetes_integration.py::test_dbt_dag_with_watcher_kubernetes PASSED [100%]
```

Example of successful run:
https://github.com/astronomer/astronomer-cosmos/actions/runs/24133899134/job/70416991590

It can be quite handy to run these tests locally to see what is going on. I usually do this by running:
```
$ hatch run tests.py3.12-3.0-1.11test-kubernetes-setup
$ hatch run tests.py3.12-3.0-1.11:test-kubernetes
```

Since we know which DAG is misbehaving, another option is to run only that specific DAG locally. I really like to do this if I want to use pdb to add break points to the code and troubleshoot what is going on:

```bash
# Export some environment variables
export CI=1
export AIRFLOW_HOME=`pwd`
export PYTHONPATH=`pwd`

# Create a symbolic link to dags
 ln -s dev/dags dags  

# Run DAG of interest
airflow dags test jaffle_shop_watcher_kubernetes
```

When I want to isolate this DAG from all the other example DAGs, I normally also add the following `.airflowignore` file to the `dags` folder, so Airflow only considers the DAG I'm troubleshooting:

```
basic_cosmos_dag_full_module_path_imports.py
basic_cosmos_dag.py
basic_cosmos_task_group_different_owners.py
basic_cosmos_task_group.py
cosmos_callback_dag.py
cosmos_manifest_example.py
cosmos_manifest_selectors_example.py
cosmos_profile_mapping.py
cosmos_seed_dag.py
cross_project_dbt_ls_dag.py
cross_project_manifest_dag.py
dbt_docs.py
example_cosmos_cleanup_dag.py
example_cosmos_dbt_build.py
example_cosmos_python_models.py
example_cosmos_sources.py
example_dbt_deps.py
example_duckdb_dag.py
example_model_version.py
example_operators.py
example_source_pruning.py
example_source_rendering.py
example_task_mapping.py
example_taskflow_operator_args.py
example_taskflow_project_config.py
example_tasks_map.py
example_tests_multiple_parents.py
example_virtualenv_mini.py
example_virtualenv.py
example_watcher.py
jaffle_shop_kubernetes.py
#jaffle_shop_watcher_kubernetes.py
performance_dag.py
simple_dag_async.py
user_defined_profile.py
```

Please, let me know if you have any questions and we can give some advice on how to run things locally, so we can merge this change.

**Comment by vricciardulli** [comment]
@tatiana thank you so much for the help! I was able to run the kubernetes tests locally and I can confirm that they don't hang anymore.

As far as I understand, the `manifest.json` for dbt project `jaffle_shop` was stale and contained 3 tests that were not actually there anymore. I remade the manifest. The tests should pass now 🤞 

I'm still not very happy that I'm using globals, so happy to hear your thoughts on that!

File: cosmos/operators/watcher_kubernetes.py [hunk 2]
```diff
@@ -67,19 +66,22 @@ def progress_callback(
...
 
 class DbtProducerWatcherKubernetesOperator(DbtBuildKubernetesOperator):
-
     template_fields: tuple[str, ...] = tuple(DbtBuildKubernetesOperator.template_fields) + ("deferrable",)
-    _process_log_line_callable: Callable[[str, dict[str, Any]], None] | None = store_dbt_resource_status_from_log
 
     def __init__(self, *args: Any, **kwargs: Any) -> None:
         task_id = kwargs.pop("task_id", "dbt_producer_watcher_operator")
+        self.tests_per_model: dict[str, list[str]] = kwargs.pop("tests_per_model", {})
+        self.test_results_per_model: dict[str, list[str]] = {}
```
[the remark this answers is not in the document]
Comment (vricciardulli) [reply cosmos/operators/watcher_kubernetes.py:85]: This was already the case before this PR, I can change it though.

**Comment by vricciardulli** [comment]
@tatiana sorry for the many messages, I was thinking that, to avoid using globals, I could instead pass the additional `kwargs` needed by `progress_callback` via `CosmosKubernetesPodManager`. Let me know if you'd like to see that implementation, no rush of course.

**Comment by vricciardulli** [comment]
@tatiana I've synced this branch with `main` and I've also removed the use of globals. Please let me know if there's anything else I should change.

File: cosmos/airflow/_override.py [hunk 3]
```diff
@@ -123,6 +134,7 @@ def consume_logs(  # noqa: C901
                                         container_name=container_name,
                                         timestamp=message_timestamp,
                                         pod=pod,
+                                        **self._callback_extra_kwargs,
```
Comment (tatiana) [inline cosmos/airflow/_override.py]: `CosmosKubernetesPodManager` now spreads `callback_extra_kwargs` into every registered callback’s `progress_callback`. Since `DbtProducerWatcherKubernetesOperator` preserves user-supplied callbacks and then appends `WatcherKubernetesCallback`, any custom callback without `**kwargs` can now receive Cosmos-only args (`tests_per_model`, `test_results_per_model`, `context`) and fail while reading pod logs. Could we scope these extra kwargs to `WatcherKubernetesCallback` only, or filter them based on the callback signature?

File: cosmos/operators/watcher_kubernetes.py [hunk 4]
```diff
@@ -72,20 +68,21 @@ def progress_callback(
...
         """
-        if "context" not in kwargs:
-            # This global variable is used to make the task context available to the K8s callback.
-            # While the callback is set during the operator initialization, the context is only created during the operator's execution.
-            kwargs["context"] = producer_task_context
-        store_dbt_resource_status_from_log(line, kwargs)
+        store_dbt_resource_status_from_log(
+            line,
+            kwargs,
+            tests_per_model=kwargs.get("tests_per_model"),
```
Comment (tatiana) [inline cosmos/operators/watcher_kubernetes.py:81]: The Kubernetes log reader can replay recent pod log lines after interruptions, so the current aggregation helper’s count-based collection can treat a duplicated terminal event for one test as if another distinct test had reported. Could we change the accumulator key by `test_unique_id` instead of appending raw statuses, so duplicate log lines are idempotent?

File: cosmos/operators/watcher_kubernetes.py [hunk 5]
```diff
@@ -121,10 +126,7 @@ def execute(self, context: Context, **kwargs: Any) -> Any:
 
         _init_xcom_backup(context)
 
-        # This global variable is used to make the task context available to the K8s callback.
-        # While the callback is set during the operator initialization, the context is only created during the operator's execution.
-        global producer_task_context
-        producer_task_context = context
+        self._context = context
```
Comment (tatiana) [inline cosmos/operators/watcher_kubernetes.py:148]: `self._context` is no initialized in `__init__`. In the normal flow, execute() sets it before `super().execute(`) triggers pod_manager, so it works - but any access to `pod_manager` before `execute` raises AttributeError. Please, add `self._context: Context | None = None` in `__init__` for robustness (and to match how `_tests_per_model/_test_results_per_model` are initialized).

**Review by tatiana — COMMENTED** [review]
Thanks for sticking with this @vricciardulli! I'm sorry for the delay.

I like the move off the globals onto `callback_extra_kwargs`, and reusing the proven `DbtTestWatcherOperator` shape (including the `split(".", 2)[2]` selector for versioned models) is very good.  Nice work on tracking down the stale `jaffle_shop` manifest.

A couple of things before we can merge:
- Please, rebase, and confirm all checks pass
- Address the comments inline.
- Update the watcher Kubernetes documentation, please

We want to ship this in Cosmos 1.15.0.  Once the rebase + the two items above are in, this should be good to go.

@johnhoran would love your thoughts on this one as well

File: cosmos/operators/watcher_kubernetes.py (hunk 5 above)
Comment (vricciardulli) [reply cosmos/operators/watcher_kubernetes.py:148]: fixed in https://github.com/astronomer/astronomer-cosmos/pull/2543/changes/f2c84c3b88d31aed651f194821aeb61f9bc1f930

File: cosmos/airflow/_override.py (hunk 3 above)
Comment (vricciardulli) [reply cosmos/airflow/_override.py]: fixed in https://github.com/astronomer/astronomer-cosmos/pull/2543/changes/8cf34d3bd371fbbe7b473f33997d1af084cd3b53

I decided to go with a class var to use as a marker. While I could have inspected the signature, I thought using a marker is more explicit and doesn't rely on python internals. Curious to know what you think.

File: cosmos/operators/watcher_kubernetes.py (hunk 4 above)
Comment (vricciardulli) [reply cosmos/operators/watcher_kubernetes.py:81]: should be fixed in https://github.com/astronomer/astronomer-cosmos/pull/2543/changes/ae12e0e9604181bcf68c63339d540180af106de9

**Comment by vricciardulli** [comment]
Hello @tatiana, thank you for the review! I've synced this branch with main, resolved all conflicts, fixed the issues you pointed out and updated the docs.

Please let me know if I missed anything:) cc @johnhoran

**Comment by tatiana** [comment]
@vricciardulli thank you very much for addresing the feedback so promptly

I'm approving and merging this PR even though I haven't tested the feature in Airflow itself. We'll be cutting Cosmos 1.15.0a1 today or tomorrow, and I want this included. We'll be validating against a few DAGs while we test 1.15.0a1, including K8s Watcher ones.

As a follow-up work, I think there is still a stale-context edge case here. Because `pod_manager` is cached, this dict captures `self._context` only when the manager is first created. If `pod_manager` is accessed before `execute()` sets `_context`, the cached manager keeps `context=None`, and later log callbacks will not be able to push the model/test status XComs.

In a follow-up PR, could we either update the cached manager's `callback_extra_kwargs["context"]` in `execute()` after line 148, or have the callback read the current context via a mutable holder/reference? It would also be good to add a regression test that accesses `op.pod_manager` before `op.execute(context=...)` and verifies that the callback kwargs use the execution context?

Thanks a lot for the amazing work here. Several community members are waiting for this feature, and I'm very excited to have it shipped in the next Cosmos release.

File: cosmos/airflow/_override.py (hunk 3 above)
Comment (tatiana) [reply cosmos/airflow/_override.py]: I like the approach, @vricciardulli , thank you!

**Review by tatiana — APPROVED** [review]
